Remove `TimeoutCancellationException` from `Publisher.stop`

lawrence-forooghian opened the pull request:
## Background

It’s not clear for callers of `Publisher.stop` what it means if it throws `TimeoutCancellationException`, or how they should respond to it. For example, do they need to try calling it again, or is the publisher now in some terminal broken state?

(The same questions also apply if the publisher throws `ConnectionException`; I’ve created #873 for addressing this.)

## What’s changed

This comes from the change suggested by @lmars in https://github.com/ably/ably-asset-tracking-android/issues/862#issuecomment-1366083580. 

We remove the operation-wide timeout parameter, and instead expose a parameter to give the caller control over only the presence leave operation.

We explain what will happen if this timeout elapses and how this affects the publisher’s online status from the point of view of subscribers.

## API considerations

From 944d88461cb36318032136adcf3b085c8e4da3d9:

> This is technically a breaking change to the publisher SDK API, both in terms of semantics and parameter name. I’ll start a discussion on the pull request to decide how we want to address this, and then update this commit message.

My change to `UPDATING.md` anticipates this just being a minor version bump. Do we think this warrants a major version bump?

## Reviewing

Please review commit-by-commit.

## Issues closed

Closes #862.

KacperKluka commented on publishing-sdk/src/main/java/com/ably/tracking/publisher/Publisher.kt:
```diff
@@ -119,13 +118,12 @@ interface Publisher {
      * Please note that calling this method will remove the notification provided by [Builder.backgroundTrackingNotificationProvider].
      *
-     * @param timeoutInMilliseconds After this duration the stopping procedure will be canceled. Default value is 30 seconds.
+     * @param presenceLeaveTimeoutInMilliseconds After this duration, the attempt to leave the Ably channel presence will be cancelled. If this timeout elapses, the remainder of the stopping procedure will still be performed, and subscribers will see the publisher go offline within 75 seconds. Default value is 30 seconds.
```
The same thing about those `75 seconds` as before 😉

lawrence-forooghian replied:
Good point! I've updated the documentation, and also updated it to make it consistent with the rest (to talk about how long it may take for subscribers to see the publisher go offline starting from the moment the method is called instead of the moment it returns).

KacperKluka reviewed (commented):
Overall looks good, just one thing to clarify 😉

lawrence-forooghian commented on UPDATING.md:
```diff
@@ -1,5 +1,13 @@
+
+It is no longer possible to specify a timeout for the entire `Publisher.stop` operation. The `timeoutInMilliseconds` parameter has been replaced by a `presenceLeaveTimeoutInMilliseconds` parameter, which allows you to specify a timeout for part of this operation.
+
+The `Publisher.stop` method now no longer throws `TimeoutCancellationException`, and guarantees that subscribers will see the publisher go offline within 75 seconds of this method returning.
```
Good point! I've updated the documentation, and also updated it to make it consistent with the other (to talk about how long it may take for subscribers to see the publisher go offline starting from the moment the method is called instead of the moment it returns).

lawrence-forooghian commented on common/src/main/java/com/ably/tracking/common/Ably.kt:
```diff
@@ -224,10 +224,15 @@ interface Ably {
      * @param presenceData The data that will be send via the presence channels.
+     * @param presenceLeaveTimeoutInMilliseconds If this value is provided, then it specifies a maximum amount of time that this method will wait for a call to [AblySdkRealtime.Presence.leave] to complete.
+     *
+     * After this timeout elapses, this method will no longer wait for the presence leave request to complete, and will proceed to perform the remainder of its work. In particular, it will still call [AblySdkRealtime.close] on the underlying Realtime object. As described by [RTN12](https://sdk.ably.com/builds/ably/specification/main/features/#RTN12), this means that whatever state the Realtime ob ...
```
Updated, thanks!

KacperKluka reviewed (approved):
LGTM 👍

lawrence-forooghian commented:
@KacperKluka @davyskiba did you have any thoughts about what kind of version bump this necessitates (perhaps Kacper you have an idea of how we've handled this sort of thing in the past)? (See "API considerations" in the description.)

KacperKluka commented:
@lawrence-forooghian IMO this particular API situation is debatable as you only change a parameter name, so this only affects users that use the named parameters in functions. Due to this I'd be fine with either bumping the minor number or bumping the major one 😉

paddybyers reviewed (approved):
I've look at your reasoning in the comments and I'm happy to approve that approach. I haven't reviewed the code itself or the tests.

QuintinWillison commented:
@lawrence-forooghian @KacperKluka I've had a look at the proposed change to the `stop` method (being the rename of the `Long` argument from `timeoutInMilliseconds` to `presenceLeaveTimeoutInMilliseconds`).

I don't believe it's correct to summarise the impact of this change as "only affects users that use the named parameters in functions": If a user doesn't use named parameters when calling this method but does pass a value to override this argument's default value then that value will now be interpreted in a different way, resulting in a change in SDK behaviour, regardless of the name.

Therefore, I think the change as currently presented is a breaking / incompatible change (`major`) whichever way we look at it.

A couple of alternative routes spring to mind that might make this easier to explain to customers when we do a release (migration guide, etc..):

1. Change the argument type entirely, perhaps something like a `StopOptions` data class that has a `presenceLeaveTimeoutInMilliseconds` member.
2. Use deprecation and allow the timeout value to be optionally injected elsewhere:
    - For the current `stop(Long = 30_000L)` method, leave the argument in place, remove the default value and deprecate it -- it becomes `stop(Long)` (deprecated)
    - add a new overload for `stop()` that takes no arguments
    - add a builder method to allow users to inject this new timeout value before they even start the publisher -- I'm pretty sure they'll know this up front, right?

Option 1 is definitely a breaking change (`major`), but option 2 could be a new feature / enhancement (`minor`).
My preference would be option 2.

jaley commented:
@lawrence-forooghian (or maybe @paddybyers or @lmars) - does the user really *need* control over this presence leave timeout? It looks a bit odd to me, as Ably presence functionality isn't really exposed to the user elsewhere in the API, it's more of implementation strategy for delivering the required experience, AFAICT? If there's no strong reason to expose the presence leave timeout, would it not be better to just do part of @QuintinWillison's suggestion (2) above:

* Add a new `stop()` overload that takes no parameters
* Deprecate `stop(Long)` and just document that the timeout parameter is now ignored and we won't throw that exception.

This isn't source-breaking, and the only behaviour change is that we'll no-longer throw an exception, which has already been agreed as an improvement from the user perspective?

paddybyers commented:
>  does the user really need control over this presence leave timeout?

I don't think so.

> would it not be better to just do part of @QuintinWillison's suggestion (2) above:

- Add a new stop() overload that takes no parameters
- Deprecate stop(Long) and just document that the timeout parameter is now ignored and we won't throw that exception.

I'm ok with that

lawrence-forooghian commented:
Before I go ahead with deprecating the ability to control the duration of the publisher `stop()` method call, I'd like to be sure that we're not removing something the client is depending on. @KacperKluka do you remember what the original motivation was for adding that timeout parameter? Was it based on some user need?

KacperKluka commented:
> do you remember what the original motivation was for adding that timeout parameter

I don't remember this exactly but I think we simply wanted to avoid freezing the SDK as the `stop` method is going to block the internal worker queue in order to correctly perform the stop operation 🤔 As far as I remember it wasn't a request from a customer but rather our idea 😉

davyskiba commented on subscribing-sdk/src/main/java/com/ably/tracking/subscriber/workerqueue/workers/StopConnectionWorker.kt:
```diff
@@ -22,7 +22,7 @@ internal class StopConnectionWorker(
         runBlocking {
             try {
-                ably.close(properties.presenceData)
+                ably.close(properties.presenceData, presenceLeaveTimeoutInMilliseconds = null)
```
What is the reasoning behind treating subscriber stop differently?

lawrence-forooghian replied:
I don't know – this operation previously had no timeout and so that's the behaviour I’ve preserved.

jaley commented:
> deprecating the ability to control the duration of the publisher stop() method call, I'd like to be sure that we're not removing something the client is depending on.

This is a fair point. Another option then would actually be to leave the function signature as it is, honour the timeout but not throw an exception if it expires, simply give up and stop blocking the caller. Was that option considered? (Sorry, I may have missed earlier discussion)

lawrence-forooghian commented:
If we believe that having any sort of timeout on `stop` is a mistake, should we just remove the argument from that method, increase the major version and call it a day? Any other kind of change (ignoring the parameter, not throwing an exception etc) is all, technically, a breaking change to the API.

lawrence-forooghian commented:
As agreed in our call this morning, I’ve removed the timeout logic and removed the corresponding parameter on `stop()`. I've written a bit about this decision in the commit message.

I’ve updated the `UPDATING.md` to explain how users can replace the timeout by using Kotlin’s `withTimeout`.

@KacperKluka @davyskiba I've added a couple of tests to give me confidence that if a user calls `stop()` from within a `withTimeout` block then the `TimeoutCancellationException` does indeed reach the user. I think that they probably come from my poor knowledge of Kotlin coroutines, so if you think them unnecessary please let me know and I'll get rid of them.

davyskiba commented on publishing-sdk/src/test/java/com/ably/tracking/publisher/DefaultPublisherTest.kt:
```diff
@@ -221,4 +225,35 @@ class DefaultPublisherTest {
+        }
+
+        // ...and the withTimeout block throws a TimeoutCancellationException.
+        Assert.assertTrue(caughtTimeoutCancellationException)
```
Please use [Expectations](https://github.com/ably/ably-asset-tracking-android/blob/main/android-test-common/src/main/java/com/ably/tracking/test/android/common/Expectations.kt) instead of assertions here

lawrence-forooghian replied:
Happy to change it, but would you mind explaining why? It seems like the main functionality of the `Expectation` class is to provide the ability to wait for some asynchronous operation to complete. But we're already waiting for the coroutine call to complete by using `runBlocking`.

davyskiba replied:
Right await is not needed here, I would use it here mostly for a more verbose messages when the test fails

lawrence-forooghian replied:
Does that mean we can stick to using assert, or is there some other reason to switch to using expectations?

davyskiba replied:
Yes, we can :)
I prefer to use something with a more informative error message here, like Expectation or Truth fluent assertion, but it is not worth keeping this pr open longer just for that

KacperKluka commented on publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/StopWorker.kt:
```diff
@@ -26,19 +24,16 @@ internal class StopWorker(
+                callbackFunction(Result.success(Unit))
             } catch (exception: ConnectionException) {
                 callbackFunction(Result.failure(exception))
             } catch (exception: TimeoutCancellationException) {
```
Since the `withTimeout` block is removed I think we no longer need to catch the `TimeoutCancellationException` 🤔

lawrence-forooghian replied:
Good point, removed.

KacperKluka reviewed (changes requested):
Looks good, just a few small things to improve 😉

lawrence-forooghian commented on publishing-sdk/src/main/java/com/ably/tracking/publisher/DefaultPublisher.kt:
```diff
@@ -112,6 +111,14 @@ constructor(
         }
     }
 
+    @Deprecated(
+        "The timeoutInMilliseconds parameter is now ignored and should not be used.",
+        replaceWith = ReplaceWith("stop()")
+    )
+    override suspend fun stop(timeoutInMilliseconds: Long) {
+        stop()
+    }
```
I added it because Android Studio was giving me this warning:

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/53756884/211378959-af97ea7f-1158-4dc5-8156-35da73d01148.png">

Should I ignore the warning, or keep it as it currently is, or suppress the warning in some other way?

lawrence-forooghian replied:
(I just tried removing it, and it actually seems to be a build error, not just a warning - try running `./gradlew check` to reproduce.)

KacperKluka replied:
Oh sorry, this must be my mistake, I was recently deprecating some things in AAT but they were regular methods, not an interface-implementation pairs 😅 It looks like in this situation the `@Deprecated` annotation needs to be added in both places 👍 Sorry for the trouble 😉

KacperKluka reviewed (approved):
Good job! LGTM 👍
